src/rgw/driver/sfs/sqlite/sqlite_users.cc

```diff
@@ -16,111 +16,117 @@
+     << user.uinfo.subusers << user.uinfo.suspended << user.uinfo.max_buckets
+     << user.uinfo.op_mask << user.uinfo.caps << user.uinfo.admin
+     << user.uinfo.system << user.uinfo.default_placement.name
+     << user.uinfo.default_placement.storage_class << user.uinfo.placement_tags
+     << user.uinfo.quota.bucket_quota << user.uinfo.temp_url_keys
+     << user.uinfo.quota.user_quota << user.uinfo.type << user.uinfo.mfa_ids
+     << nullptr << user.user_attrs << user.user_version.ver
+     << user.user_version.tag;
+  _store_access_keys(user);
 }
 
 void SQLiteUsers::remove_user(const std::string& userid) const {
-  auto storage = conn->get_storage();
-  _remove_access_keys(storage, userid);
-  storage->remove<DBUser>(userid);
+  dbapi::sqlite::database db = conn->get();
+  _remove_access_keys(userid);
+  db << R"sql(
+    DELETE FROM users
+    WHERE user_id = ?;)sql"
+     << userid;
 }
 
-template <class... Args>
-std::vector<DBOPUserInfo> SQLiteUsers::get_users_by(Args... args) const {
-  std::vector<DBOPUserInfo> users_return;
-  auto storage = conn->get_storage();
-  auto users = storage->get_all<DBUser>(args...);
-  for (auto& user : users) {
-    users_return.push_back(get_rgw_user(user));
-  }
-  return users_return;
-}
-
-void SQLiteUsers::_store_access_keys(
-    StorageRef storage, const DBOPUserInfo& user
-) const {
+void SQLiteUsers::_store_access_keys(const DBOPUserInfo& user) const {
   // remove existing keys for the user (in case any of them had changed)
-  _remove_access_keys(storage, user.uinfo.user_id.id);
+  _remove_access_keys(user.uinfo.user_id.id);
+  dbapi::sqlite::database db = conn->get();
   for (auto const& key : user.uinfo.access_keys) {
-    DBAccessKey db_key;
-    db_key.access_key = key.first;
-    db_key.user_id = user.uinfo.user_id.id;
-    storage->insert(db_key);
+    db << R"sql(INSERT INTO access_keys (access_key, user_id) VALUES (?,?);)sql"
+       << key.first << user.uinfo.user_id.id;
   }
 }
 
-void SQLiteUsers::_remove_access_keys(
-    StorageRef storage, const std::string& userid
-) const {
-  storage->remove_all<DBAccessKey>(where(c(&DBAccessKey::user_id) = userid));
+void SQLiteUsers::_remove_access_keys(const std::string& userid) const {
+  dbapi::sqlite::database db = conn->get();
+  db << R"sql(
+    DELETE FROM access_keys
+    WHERE user_id = ?;)sql"
+     << userid;
 }
 
 std::optional<std::string> SQLiteUsers::_get_user_id_by_access_key(
-    StorageRef storage, const std::string& key
+    const std::string& key
 ) const {
-  auto keys =
-      storage->get_all<DBAccessKey>(where(c(&DBAccessKey::access_key) = key));
-  std::optional<std::string> ret_value;
-  if (keys.size() > 0) {
+  std::optional<std::string> ret;
+  dbapi::sqlite::database db = conn->get();
+  auto rows = db << R"sql(
+    SELECT user_id FROM access_keys
+    WHERE access_key = ?;)sql"
+                 << key;
+  for (std::tuple<std::string> row : rows) {
```

Comment:
I'd avoid using a `for` in this way for readability, if we are interested in the first record, there should be a more straightforward way to obtain that.
Also, same considerations on moves,  maybe you can use the && notation here if row can be moved and use std::move(row) in the subsequent assignment

Reply:
> I'm afraid this library is not offering direct access to rows.

I think this is more than reasonable, otherwise it would mean the library would load in memory an arbitrary amount of rows, I really hope the library is implementing an iterator like behavior :) .

I think enhanced for in C++ is just a sugar-syntax wrapper using an iterator taken from the collection you are iterating on.
If this were the case should not be difficult accessing the first element with a row.begin() statement.
Anyway, as you prefer, saying this just because I don't like seeing a for used in this way, but it is just personal taste.

Reply:
oh, do you mean by doing this?
```c++
auto row = *rows.begin(); 
```
I tried that, but it didn't work as expected. 

I tried doing this:
```c++
if (rows.begin() != rows.end()) {
  std::tuple<std::string> val = *rows.begin();
}
```

The above didn't work because calling `begin` changes the state internally and the next time we call `begin` value is not returned as expected. I gave up because although I dislike the `for` thing there's a lot to do with the port.

This seems to work fine, so I've changed the code to do it this way: 
```c++
auto iter = rows.begin();
if (iter != rows.end()) {
  std::tuple<std::string> row = *iter;
  ...
}
```

Reply:
Yup, as said, this is matter of personal tastes, not really an issue ;).
Thanks!